zephyr/wrapper.c

```diff
@@ -74,6 +79,14 @@ static void *heap_alloc_aligned(struct k_heap *h, size_t align, size_t bytes)
 
 	k_spinlock_key_t key = k_spin_lock(&h->lock);
 
+	/*
+	 * To ensure all cores have a coherent view of the heap,
+	 * when accessing the heap, flush and invalidate the whole
+	 * cache while holding the spin lock.
+	 */
+	if (h == &sof_heap)
+		z_xtensa_cache_flush_inv_all();
```

Comment:
firstly, wouldn't synchronising only heap's cache be enough? BTW, have we tried per-core non-shared / cacheable heaps?

Reply:
@lyakh  The exact location of the metadata chunks is implementation specific (of sys_heap) and any alloc/free may access arbitrary chunk headers, so we'd have to go over the whole heap area to flush and invalidate. I think just flushing all is easier here.

I did not try per-core heaps. Sizing the heaps (for per core heaps) seems like a really tough problem. Memory is a precious resource and we'd have to make it fairly dynamic to ensure one can run topologies where core memory usage is not balanced between the cores. Current Zephyr heap doesn't allow resizing (arguably not a very typical feature).